<?xml version="1.0"?>

<launch>
    <arg name="map" default="$(find full_coverage_path_planner)/maps/grid.yaml"/>
    <arg name="coverage_area_offset" default="-2.5 -2.5 0 0 0 0"/>
    <arg name="coverage_area_size_x" default="10"/>
    <arg name="coverage_area_size_y" default="10"/>
    <arg name="target_x_vel" default="0.5"/>
    <arg name="target_yaw_vel" default="0.4"/>
    <arg name="robot_radius" default="0.3"/>
    <arg name="tool_radius" default="0.3"/>
    <arg name="rviz" default="true"/>

    <!-- Mobile robot simulator -->
    <node pkg="mobile_robot_simulator" type="mobile_robot_simulator_node" name="mobile_robot_simulator" output="screen">
        <param name="publish_map_transform" value="true"/>
        <param name="publish_rate" value="10.0"/>
        <param name="velocity_topic" value="/cmd_vel"/>
        <param name="odometry_topic" value="/odom"/>
    </node>

    <!--Run the map server-->
    <node name="grid_server" pkg="map_server" type="map_server" args="$(arg map)">
        <param name="frame_id" value="map"/>
    </node>

    <!--Tracking_pid cannot just accept a nav_msgs/Path, it can only go to a PoseStamped,
       so the path_interpolator drags a PoseStamped over a Path at a given speed-->
    <node name="interpolator" pkg="tracking_pid" type="path_interpolator">
        <param name="target_x_vel" value="$(arg target_x_vel)"/>
        <param name="target_yaw_vel" value="$(arg target_yaw_vel)"/>
        <!-- <remap from="path" to="/move_base/BoustrophedonSTC/plan"/> -->
    </node>

    <!--Tracking_pid tries to get the robot as close to it's goal point as possible-->
    <node name="controller" pkg="tracking_pid" type="controller" output="screen">
        <remap from="move_base/cmd_vel" to="/cmd_vel"/>
        <remap from="local_trajectory" to="trajectory"/>
        <param name="controller_debug_enabled" value="True"/>
        <param name="track_base_link" value="true"/>
        <param name="l" value="0.5"/>
        <param name="Ki_long" value="0.0"/>
        <param name="Kp_long" value="2.0"/>
        <param name="Kd_long" value="0.5"/>
        <param name="Ki_lat" value="0.0"/>
        <param name="Kp_lat" value="4.0"/>
        <param name="Kd_lat" value="0.3"/>
    </node>

    <!-- Launch coverage progress tracking -->
    <node pkg="tf" type="static_transform_publisher" name="map_to_coveragemap" args="$(arg coverage_area_offset) map coverage_map 100" />
    <node pkg="full_coverage_path_planner" type="coverage_progress" name="coverage_progress">
        <param name="~target_area/x" value="$(arg coverage_area_size_x)" />
        <param name="~target_area/y" value="$(arg coverage_area_size_y)" />
        <param name="~coverage_radius" value="$(arg tool_radius)" />
        <remap from="reset" to="coverage_progress/reset" />
        <param name="~map_frame" value="/coverage_map"/>
    </node>

    <!-- rviz -->
    <node if="$(eval rviz)" name="rviz" pkg="rviz" type="rviz" args="-d $(find full_coverage_path_planner)/test/full_coverage_path_planner/fcpp.rviz" />
</launch>
